# 01 — Coleta Atlas (GET) — documentação da disciplina

**Projeto:** Preditor de Falhas ML (Grupo 16)
**Disciplinas:** ED2 + Redes + APS
**Card:** S1.2 — GET `/results/` (camada raw)

> Este notebook é a **documentação / evidência** pedida pela professora.
> **Não** é o pipeline de produção.
> As funções de acesso à API ficam **neste notebook** (seção 3). O pacote
> `preditor_de_falhas_ml` desta fase foi removido na refatoração; hoje o projeto
> lê as medições do BigQuery (`src/preditor/`, ver `README.md`).
> HTTP e autenticação (`requests.request` + `Authorization`) vivem em `_request`.
> **Não** criamos medição aqui (sem POST). Só lemos `/results/` de um `msm_id` já existente.

**Objetivo:** consumir resultados de uma medição do RIPE Atlas, receber o JSON,
exibi-lo em um `DataFrame` e gravar a camada `raw` em `data/raw/` (equivalente
à pasta `raw` do template da disciplina; não usamos Google Drive).

> Esta etapa cobre **somente a ingestão e o armazenamento da camada `raw`**.
> Não agregue janelas, não crie variáveis, não rotule `status_real` e não treine
> modelos aqui.

## Entrega mínima

- Justificar as decisões da equipe sobre os parâmetros da consulta;
- Executar uma requisição válida à API via `fetch_measurement_results` (definida na seção 3);
- Exibir os dados em um `DataFrame`;
- Salvar o JSON original da resposta;
- Salvar os registros em **CSV** em `data/raw/`;
- Registrar evidências da coleta: URL, parâmetros, horário, quantidade de registros e nomes dos arquivos.

## 1. Identificação da equipe

| Integrante | Papel nesta entrega |
|---|---|
| Guilherme Leite Tavares | Arquitetura de dados, Quadro Kanban, especificação e revisão |
| Alexandre Tiago de Oliveira | N/A |
| Ingrid Ferreira de Sousa | N/A |
| Kauan Garcia Dias de Oliveira | Revisão do notebook / aprovação do PR |
| Lucas Eduardo Malachias Bagatela | N/A |
| Stephanie Vitoria Bessa dos Santos | Revisão / aprovação de PRs |

- **Turma:** Ciência da Computação 3 semestre e 4 semestre
- **Data da coleta (demo no notebook):** 2026-09-14
- **Repo:** https://github.com/g-tavares14/preditor-de-falhas-ml


## 2. Decisões da equipe

| Decisão | Escolha | Por quê |
|---|---|---|
| API | RIPE Atlas | Fonte definida do projeto; não usamos outra API |
| Verbo nesta fase | **GET** `/measurements/{msm_id}/results/` | Coleta/ingestão; POST (criar medição) foi setup (S1.6/S1.8), não esta demo |
| Tipo da medição | ping | Tipo principal do dataset do preditor; traceroute entra no POST (notebook 02) |
| `msm_id` da demo | `210928969` (ping → 94.140.14.14 AdGuard) | Um dos 8 IDs reais do hub; série periódica já criada |
| Janela | 2026-09-13 12:00 UTC → 2026-09-14 12:00 UTC | Período em que a série S1.8 gerou histórico (hoje a série está Stopped; GET do histórico permanece) |
| Onde está o código | Funções na seção 3 deste notebook | O pacote desta fase foi removido; as demais células só chamam essas funções |
| Papel deste notebook | Documentação + evidência da disciplina | Não é o pipeline AWS |
| Formato tabular | **CSV** | Exigência da fase; fácil de abrir na correção |
| Pasta | `data/raw/` (raiz do repo) | Camada raw do repo (equivale à pasta `raw` pedida no template) |

## 3. Bibliotecas e funções de acesso à API

Na raiz do repositório: `uv sync --group notebook` (não usamos `pip` nem o template Colab).
As versões das bibliotecas ficam fixadas em `pyproject.toml` e `uv.lock`.
`requests` só aparece dentro de `_request`; as demais células chamam as funções abaixo.
Defina `RIPE_ATLAS_API_KEY` no ambiente ou `.env` (**não** commitado).

- `_request` — faz a chamada HTTP com a chave no cabeçalho `Authorization: Key …`,
  imprime o status HTTP e **interrompe** se ele não for 2xx (`raise_for_status`);
- `fetch_measurement_results` — GET `/measurements/{msm_id}/results/`; recusa
  qualquer resposta que não seja uma lista de resultados e devolve **a lista
  original da API** (o bruto que vai para o `.json`) e o `DataFrame` feito dela.

In [1]:
import json
import os
from datetime import UTC, datetime
from pathlib import Path

import pandas as pd
import requests

API = "https://atlas.ripe.net/api/v2/"


def _request(method, url, api_key, **kwargs):
    """Chamada HTTP autenticada ao RIPE Atlas; devolve o JSON da resposta.

    Erro HTTP (401, 403, 404, 5xx...) interrompe aqui, em vez de virar DataFrame.
    """
    response = requests.request(
        method,
        url,
        headers={"Authorization": f"Key {api_key}", "Accept": "application/json"},
        timeout=30,
        **kwargs,
    )
    print("Status HTTP:", response.status_code)
    response.raise_for_status()
    return response.json()


def fetch_measurement_results(api_key, msm_id, *, start=None, stop=None):
    """GET /measurements/{msm_id}/results/. Sem POST, sem features: só o bruto.

    Devolve (results, DataFrame): `results` é a lista exatamente como a API mandou.
    """
    params = {k: v for k, v in {"start": start, "stop": stop}.items() if v is not None}
    results = _request("GET", f"{API}measurements/{msm_id}/results/", api_key, params=params)
    if not isinstance(results, list):
        raise ValueError(f"Resposta inesperada (esperava lista de resultados): {str(results)[:300]}")
    return results, pd.DataFrame(results)


print("Bibliotecas e funções carregadas.")
print("HTTP: fetch_measurement_results -> _request (requests só dentro de _request).")

Bibliotecas e funções carregadas.
HTTP: fetch_measurement_results -> _request (requests só dentro de _request).


## 4. Preparar a pasta `raw`

O template da disciplina usa Google Drive (`ripe_atlas/raw`). Aqui a camada
equivalente é `data/raw/` na raiz do repositório. A subpasta **precisa** chamar-se `raw`.
Os arquivos gerados estão no `.gitignore` (exceto `.gitkeep`).

**Onde executar:** abra o notebook a partir de `notebooks/` (o padrão do Jupyter/VS Code).
A célula abaixo localiza a raiz do repositório pelo `pyproject.toml`, então a pasta
`data/raw/` é a mesma também se o kernel for iniciado na raiz.

In [2]:
RAIZ = Path.cwd() if (Path.cwd() / "pyproject.toml").exists() else Path.cwd().parent
assert (RAIZ / "pyproject.toml").exists(), "Execute a partir de notebooks/ ou da raiz do repositório."
PASTA_RAW = RAIZ / "data" / "raw"
PASTA_RAW.mkdir(parents=True, exist_ok=True)
print("Arquivos raw serão salvos em:", PASTA_RAW)

Arquivos raw serão salvos em: /Users/gtavares/preditor-de-falhas-ml/data/raw


## 5. Definir os parâmetros da consulta

Os parâmetros de `fetch_measurement_results`:

- `msm_id` — id de uma medição Atlas já existente (sem POST nesta demo);
- `start` / `stop` — janela Unix UTC;
- `FORMATO` — tabular da disciplina (`csv`);
- `GRAVAR_RAW` — nesta entrega fica `True` (evidência obrigatória).

Na raiz: `uv sync`. A chave `RIPE_ATLAS_API_KEY` vem do ambiente (ou `.env`).

In [3]:
msm_id = 210928969
start = int(datetime(2026, 9, 13, 12, 0, tzinfo=UTC).timestamp())
stop = int(datetime(2026, 9, 14, 12, 0, tzinfo=UTC).timestamp())
GRAVAR_RAW = True
FORMATO = "csv"

URL = f"https://atlas.ripe.net/api/v2/measurements/{msm_id}/results/"
PARAMETROS = {"start": start, "stop": stop}

api_key = os.environ.get("RIPE_ATLAS_API_KEY", "")

assert isinstance(msm_id, int) and msm_id > 0
assert stop > start
assert FORMATO == "csv"
assert GRAVAR_RAW

print("URL:", URL)
print("Início (UTC):", datetime.fromtimestamp(start, tz=UTC).isoformat())
print("Fim (UTC):", datetime.fromtimestamp(stop, tz=UTC).isoformat())
print("Parâmetros:", PARAMETROS)
print("GRAVAR_RAW:", GRAVAR_RAW, "FORMATO:", FORMATO)
print("Chave definida:", bool(api_key))


URL: https://atlas.ripe.net/api/v2/measurements/210928969/results/
Início (UTC): 2026-09-13T12:00:00+00:00
Fim (UTC): 2026-09-14T12:00:00+00:00
Parâmetros: {'start': 1789300800, 'stop': 1789387200}
GRAVAR_RAW: True FORMATO: csv
Chave definida: True


## 6. Consumir a API e obter o JSON

A célula abaixo **chama** `fetch_measurement_results` (definida na seção 3).
`requests` só aparece dentro de `_request`; **não** há POST nesta etapa.

O HTTP (timeout de 30 s, `Authorization: Key …`) fica em `_request`, que imprime o
status e para em qualquer resposta que não seja 2xx. Sucesso nesta demo = HTTP 200
e uma lista com pelo menos um registro. Se vier vazia, revise `msm_id` ou a janela.

In [4]:
if not api_key:
    raise RuntimeError(
        "RIPE_ATLAS_API_KEY não está definida (ambiente ou ..ENV, não commitado)."
    )

instante_requisicao_utc = datetime.now(tz=UTC)
dados_json, df_raw = fetch_measurement_results(
    api_key,
    msm_id,
    start=start,
    stop=stop,
)

if df_raw.empty:
    raise ValueError(
        "A consulta não retornou registros. Reveja o msm_id ou o período."
    )

print("Função:", "fetch_measurement_results")
print("Quantidade de registros recebidos:", len(dados_json))
print("Tipo do objeto principal:", type(dados_json).__name__)
print("Coletado em (UTC):", instante_requisicao_utc.isoformat())

Status HTTP: 200
Função: fetch_measurement_results
Quantidade de registros recebidos: 569
Tipo do objeto principal: list
Coletado em (UTC): 2026-10-05T00:31:18.912854+00:00


## 7. Inspecionar uma pequena amostra do JSON

Exiba somente o primeiro registro para compreender a estrutura sem poluir o
notebook. A resposta completa será preservada em arquivo. O registro vem da lista
original da API (não do DataFrame); campos aninhados como `result[]` do ping
continuam como objetos.

In [5]:
print(json.dumps(dados_json[0], indent=2, ensure_ascii=False, default=str)[:5000])


{
  "fw": 5120,
  "mver": "2.6.4",
  "lts": 19,
  "dst_name": "94.140.14.14",
  "af": 4,
  "dst_addr": "94.140.14.14",
  "src_addr": "172.29.1.2",
  "proto": "ICMP",
  "ttl": 51,
  "size": 64,
  "result": [
    {
      "rtt": 3.977473
    },
    {
      "rtt": 3.839475
    },
    {
      "rtt": 3.924534
    },
    {
      "rtt": 3.934877
    },
    {
      "rtt": 4.076464
    },
    {
      "rtt": 3.874219
    },
    {
      "rtt": 4.306371
    },
    {
      "rtt": 4.731335
    }
  ],
  "dup": 0,
  "rcvd": 8,
  "sent": 8,
  "min": 3.839475,
  "max": 4.731335,
  "avg": 4.0830935,
  "msm_id": 210928969,
  "prb_id": 1016697,
  "timestamp": 1789344087,
  "msm_name": "Ping",
  "from": "177.136.13.34",
  "type": "ping",
  "group_id": 210928969,
  "step": 300,
  "stored_timestamp": 1789344169
}


## 8. Transformar o JSON em `DataFrame`

O template da disciplina usa `pd.json_normalize`. Aqui `fetch_measurement_results`
já devolve `pd.DataFrame(results)`. Não removemos nem agregamos registros.

In [6]:
print("Dimensões do DataFrame (linhas, colunas):", df_raw.shape)
display(df_raw.head())


Dimensões do DataFrame (linhas, colunas): (569, 26)


,fw,mver,lts,dst_name,af,dst_addr,src_addr,proto,ttl,size,...,avg,msm_id,prb_id,timestamp,msm_name,from,type,group_id,step,stored_timestamp
0,5120,2.6.4,19,94.140.14.14,4,94.140.14.14,172.29.1.2,ICMP,51.0,64,...,4.083094,210928969,1016697,1789344087,Ping,177.136.13.34,ping,210928969,300,1789344169
1,5120,2.6.4,40,94.140.14.14,4,94.140.14.14,172.29.1.2,ICMP,51.0,64,...,3.924984,210928969,1016697,1789344394,Ping,177.136.13.34,ping,210928969,300,1789344484
2,5120,2.6.4,35,94.140.14.14,4,94.140.14.14,172.29.1.2,ICMP,51.0,64,...,3.863244,210928969,1016697,1789344699,Ping,177.136.13.34,ping,210928969,300,1789344729
3,5120,2.6.4,34,94.140.14.14,4,94.140.14.14,172.29.1.2,ICMP,51.0,64,...,6.197212,210928969,1016697,1789344995,Ping,177.136.13.34,ping,210928969,300,1789345075
4,5120,2.6.4,37,94.140.14.14,4,94.140.14.14,172.29.1.2,ICMP,51.0,64,...,10.562327,210928969,1016697,1789345289,Ping,177.136.13.34,ping,210928969,300,1789345373


In [7]:
# Conferência básica das colunas e tipos identificados.
df_raw.info()


<class 'pandas.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 26 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   fw                569 non-null    int64  
 1   mver              569 non-null    str    
 2   lts               569 non-null    int64  
 3   dst_name          569 non-null    str    
 4   af                569 non-null    int64  
 5   dst_addr          569 non-null    str    
 6   src_addr          569 non-null    str    
 7   proto             569 non-null    str    
 8   ttl               568 non-null    float64
 9   size              569 non-null    int64  
 10  result            569 non-null    object 
 11  dup               569 non-null    int64  
 12  rcvd              569 non-null    int64  
 13  sent              569 non-null    int64  
 14  min               569 non-null    float64
 15  max               569 non-null    float64
 16  avg               569 non-null    float64
 17  msm_id  

## 8.1 Validação mínima e cobertura

Cada linha é **uma rajada de ping de uma sonda**: identificada por
`msm_id + prb_id + timestamp`. Conferimos, sem alterar o bruto:

- se a chave identifica cada linha (duplicatas);
- `sent` / `rcvd` por rajada e quantas rajadas foram **timeout total**
  (`rcvd = 0`, o Atlas manda `avg = -1`), separadas das que têm RTT;
- pings perdidos dentro de rajadas que tiveram resposta (`{"x": "*"}` em `result`);
- a **cobertura**: quantas sondas e quantas rajadas cada uma tem na janela,
  e o primeiro/último instante medido.

In [8]:
chave = ["msm_id", "prb_id", "timestamp"]
assert set(chave) <= set(df_raw.columns), "Colunas da chave ausentes na resposta."
assert (df_raw["msm_id"] == msm_id).all(), "Há registros de outro msm_id."
duplicatas = int(df_raw.duplicated(subset=chave).sum())

timeout_total = df_raw["rcvd"] == 0
pings_perdidos = int((df_raw["sent"] - df_raw["rcvd"]).sum())

print("Registros:", len(df_raw), "| duplicatas pela chave:", duplicatas)
print("Pings enviados:", int(df_raw["sent"].sum()), "| recebidos:", int(df_raw["rcvd"].sum()),
      "| perdidos:", pings_perdidos)
print("Rajadas com timeout total (rcvd = 0, avg = -1):", int(timeout_total.sum()))
print("Rajadas com RTT válido (avg > 0):", int((df_raw["avg"] > 0).sum()))
print("Coerência: avg = -1 exatamente quando rcvd = 0:",
      bool(((df_raw["avg"] == -1) == timeout_total).all()))
print("Nulos por coluna (só as que têm):", df_raw.isna().sum()[lambda s: s > 0].to_dict())

instantes = pd.to_datetime(df_raw["timestamp"], unit="s", utc=True)
print("Primeira medição:", instantes.min().isoformat(), "| última:", instantes.max().isoformat())

cobertura = (
    df_raw.assign(timeout_total=timeout_total)
    .groupby("prb_id")
    .agg(rajadas=("timestamp", "size"), timeouts=("timeout_total", "sum"))
    .sort_values("rajadas", ascending=False)
)
print("Sondas na janela:", len(cobertura))
display(cobertura.describe().loc[["min", "50%", "max"]])
display(cobertura.head(10))

Registros: 569 | duplicatas pela chave: 0
Pings enviados: 4547 | recebidos: 4538 | perdidos: 9
Rajadas com timeout total (rcvd = 0, avg = -1): 1
Rajadas com RTT válido (avg > 0): 568
Coerência: avg = -1 exatamente quando rcvd = 0: True
Nulos por coluna (só as que têm): {'ttl': 1}
Primeira medição: 2026-09-13T12:11:27+00:00 | última: 2026-09-14T11:56:38+00:00
Sondas na janela: 2


,rajadas,timeouts
min,283.0,0.0
50%,284.5,0.5
max,286.0,1.0


,rajadas,timeouts
prb_id,,
1016697,286,0
1016344,283,1


## 9. Salvar os dados na pasta `raw`

Serão gerados:

1. Um arquivo `.json` com **a lista original devolvida pela API** (o bruto oficial,
   sem passar pelo DataFrame);
2. Um arquivo `.csv` com o conteúdo exibido no `DataFrame`;
3. Um arquivo de metadados `.json` com as evidências da coleta.

O timestamp UTC no nome evita sobrescrever coletas anteriores: cada execução
gera uma tríade nova, e nenhuma é concatenada a outra.
Isto é só evidência da disciplina.

In [9]:
assert GRAVAR_RAW, "Gravar dados brutos é obrigatório nesta entrega."
assert len(df_raw) > 0, "Não há dados para gravar."
assert FORMATO == "csv"

stamp = instante_requisicao_utc.strftime("%Y%m%dT%H%M%SZ")
nome_base = f"ripe_atlas_m{msm_id}_{stamp}"

caminho_json = PASTA_RAW / f"{nome_base}.json"
caminho_tabela = PASTA_RAW / f"{nome_base}.csv"
caminho_metadados = PASTA_RAW / f"{nome_base}_metadata.json"

caminho_json.write_text(
    json.dumps(dados_json, ensure_ascii=False, indent=2, default=str),
    encoding="utf-8",
)
df_raw.to_csv(caminho_tabela, index=False, encoding="utf-8")

metadados = {
    "measurement_id": msm_id,
    "url": URL,
    "parametros": PARAMETROS,
    "inicio_utc": datetime.fromtimestamp(start, tz=UTC).isoformat(),
    "fim_utc": datetime.fromtimestamp(stop, tz=UTC).isoformat(),
    "coletado_em_utc": instante_requisicao_utc.isoformat(),
    "funcao": "fetch_measurement_results",
    "http": "_request (GET /api/v2/measurements/{msm_id}/results/)",
    "quantidade_registros": int(len(df_raw)),
    "quantidade_colunas_dataframe": int(len(df_raw.columns)),
    "arquivo_json": caminho_json.name,
    "arquivo_tabelar": caminho_tabela.name,
    "duplicatas_pela_chave": duplicatas,
    "rajadas_timeout_total": int(timeout_total.sum()),
    "sondas": int(len(cobertura)),
    "arquivo_json_e_bruto_original_da_api": True,
    "nota": "Demo notebook 01; funções definidas no próprio notebook",
}

caminho_metadados.write_text(
    json.dumps(metadados, ensure_ascii=False, indent=2),
    encoding="utf-8",
)

{
    "json": str(caminho_json),
    "csv": str(caminho_tabela),
    "metadata": str(caminho_metadados),
}

{'json': '/Users/gtavares/preditor-de-falhas-ml/data/raw/ripe_atlas_m210928969_20261005T003118Z.json',
 'csv': '/Users/gtavares/preditor-de-falhas-ml/data/raw/ripe_atlas_m210928969_20261005T003118Z.csv',
 'metadata': '/Users/gtavares/preditor-de-falhas-ml/data/raw/ripe_atlas_m210928969_20261005T003118Z_metadata.json'}

## 10. Verificar se a entrega foi salva

A célula abaixo verifica a existência dos arquivos da tríade da disciplina
(JSON + CSV + metadata), lista tamanhos e confere que o JSON gravado é igual
à lista recebida da API.

In [10]:
assert caminho_json.exists(), f"JSON ausente: {caminho_json}"
assert caminho_tabela.exists(), f"CSV ausente: {caminho_tabela}"
assert caminho_metadados.exists(), f"Metadados ausentes: {caminho_metadados}"
assert json.loads(caminho_json.read_text(encoding="utf-8")) == dados_json, "JSON gravado difere da resposta."

arquivos_gerados = pd.DataFrame(
    {
        "arquivo": [caminho_json.name, caminho_tabela.name, caminho_metadados.name],
        "tamanho_bytes": [
            caminho_json.stat().st_size,
            caminho_tabela.stat().st_size,
            caminho_metadados.stat().st_size,
        ],
    }
)
display(arquivos_gerados)
print("Demo oficial (stamp):", stamp)
print("Coleta concluída e salva na pasta raw.")

,arquivo,tamanho_bytes
0,ripe_atlas_m210928969_20261005T003118Z.json,521605
1,ripe_atlas_m210928969_20261005T003118Z.csv,195280
2,ripe_atlas_m210928969_20261005T003118Z_metadat...,826


Demo oficial (stamp): 20261005T003118Z
Coleta concluída e salva na pasta raw.


**Observação**

Evidências aceitas:

1. histórico de edição de documento compartilhado;
2. commit ou pull request no GitHub;
3. arquivo ou trecho de código produzido;
4. relatório, planilha, diagrama, apresentação ou rascunho;
5. registro de tarefa no Trello, GitHub Projects ou ferramenta semelhante;
6. registro de testes realizados;
7. print de reunião, conversa ou e-mail relacionado à atividade;
8. outro material que demonstre claramente a contribuição individual.

A evidência deverá identificar o aluno e estar relacionada à atividade declarada.
Sempre que possível, apresente materiais com data, autoria ou histórico de edição.
Verifique se todos os links estão acessíveis. Trabalhos realizados em conjunto
também devem indicar a contribuição específica de cada integrante. Respostas
genéricas, como “ajudei no trabalho”, “fiz a pesquisa” ou “participei do código”,
não serão consideradas suficientes. Prints de conversas podem complementar a
comprovação, mas não devem ser a única evidência quando houver um produto
verificável. A ausência de descrição ou de evidência poderá impedir a validação
da contribuição individual.


## 11. Checklist da equipe

- [x] Identificamos os integrantes e a turma;
- [x] Justificamos o ID da medição, o tipo (ping) e o intervalo escolhido;
- [x] A requisição foi um GET válido via `fetch_measurement_results` (HTTP em `_request`);
- [x] O JSON / DataFrame retornou pelo menos um registro;
- [x] Exibimos o `DataFrame` no notebook (`head` / `info`);
- [x] Conferimos status HTTP, duplicatas, `sent`/`rcvd`, timeouts e cobertura por sonda;
- [x] Salvamos o JSON original (lista da API) em `data/raw/`;
- [x] Salvamos o DataFrame em CSV em `data/raw/`;
- [x] Mantivemos o arquivo de metadados da coleta (nome com timestamp UTC);
- [x] Não realizamos limpeza, agregação, rotulação ou treinamento nesta etapa;
- [x] `requests` só dentro de `_request`; sem POST nesta etapa;
- [x] Claro para a banca: **notebook = doc da primeira fase (coleta via API)**.

### Registro final da equipe

Coletamos resultados de ping da medição `210928969` (AdGuard `94.140.14.14`, um
dos 8 `msm_id` do hub) na janela 2026-09-13 12:00 UTC → 2026-09-14 12:00 UTC,
período em que a série S1.8 gerou histórico. O GET foi feito por
`fetch_measurement_results`, definida no próprio notebook. Volume e cobertura
(seções 6, 8.1 e 10): HTTP 200, **569 registros** (rajadas de 8 pings) de **2 sondas**
(1016697: 286 rajadas; 1016344: 283, ≈ 1 a cada 5 min, `step` = 300 s), da
primeira medição às 12:11:27 à última às 11:56:38 UTC. **0** duplicatas pela chave
`msm_id + prb_id + timestamp`; 4.547 pings enviados, 4.538 recebidos (9 perdidos);
**1** rajada com timeout total (`rcvd = 0`, `avg = -1`). Demo oficial = tríade com
stamp **20261005T003118Z** em `data/raw/` (coletada em 2026-10-05 00:31 UTC). CSV é o formato tabular desta fase; JSON e metadata fecham a
tríade de evidência. Sem agregação e sem `status_real` aqui.